# Data import 

In [1]:
TRAIN_FILE = "data/TrainOnMe.csv"
EVALUATE_FILE = "data/EvaluateOnMe.csv"

In [2]:
import pandas as pd
import numpy as np

# DATA VISUALISATION

df = pd.read_csv(TRAIN_FILE)
df = df.drop(620) #We remove the line where there is no datapoint but a sentence, I saw it :)
df = df.iloc[:, 1:] #We remove the first column with the number of each row, it is useless !
df.head()

y        x1       x2         x3       x4         x5        x6  \
0     Boborg  99.88558 -0.78795  -96.54505 -0.63323  199.67072 -89.64438   
1   Jorgsuto  98.62823 -0.34918  -99.47730 -0.63830  198.40234 -90.70851   
2  Atsutobob  98.07893 -0.29079  -98.13996 -0.68610  197.88585 -89.32786   
3  Atsutobob  99.62099 -0.42186  -99.17242 -0.59933  199.39653 -88.84929   
4     Boborg  99.96833  0.04647 -105.15302 -0.64502  199.75719 -90.55073   

                 x7       x8       x9       x10      x11   x12        x13  
0  Slängpolskorgris  0.10388  1.36981  11.09455 -3.49487  True  451.15538  
1         Hambogris  0.62187 -2.31829   8.87321  1.26229  True  443.40251  
2         Hambogris -0.63982  1.15431  11.86206  2.73511  True  441.32469  
3  Slängpolskorgris -1.15077  2.16860   9.87663  2.93509  True  448.51875  
4  Slängpolskorgris -0.51528 -1.49196  11.82138 -3.21950  True  447.26516

# Cleaning and pre-analyzing the data

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 0 to 1000
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   y       1000 non-null   object 
 1   x1      1000 non-null   object 
 2   x2      1000 non-null   float64
 3   x3      1000 non-null   float64
 4   x4      999 non-null    float64
 5   x5      1000 non-null   float64
 6   x6      1000 non-null   float64
 7   x7      999 non-null    object 
 8   x8      1000 non-null   float64
 9   x9      1000 non-null   float64
 10  x10     1000 non-null   float64
 11  x11     1000 non-null   float64
 12  x12     1000 non-null   object 
 13  x13     1000 non-null   float64
dtypes: float64(10), object(4)
memory usage: 117.2+ KB


Are there duplicates ?

In [4]:
df.shape

(1000, 14)

In [5]:
df.drop_duplicates().shape

(1000, 14)

There are no duplicates !

In [6]:
df.isnull().sum()

y      0
x1     0
x2     0
x3     0
x4     1
x5     0
x6     0
x7     1
x8     0
x9     0
x10    0
x11    0
x12    0
x13    0
dtype: int64

But there are 2 null values, we delete them !

In [7]:
df_nonnull = df.dropna()
df_nonnull.shape

(998, 14)

Then we have to see how are the columns (their type, their mean, their standard deviation, etc...)

In [8]:
df_nonnull.describe(include="all")

y        x1          x2          x3          x4          x5  \
count      998       998  998.000000  998.000000  998.000000  998.000000   
unique       6       996         NaN         NaN         NaN         NaN   
top     Boborg  99.88558         NaN         NaN         NaN         NaN   
freq       420         2         NaN         NaN         NaN         NaN   
mean       NaN       NaN   -0.007644  -99.946945   -0.653185  199.769720   
std        NaN       NaN    0.730013    3.105700    0.031579    1.003122   
min        NaN       NaN   -2.457140 -111.018930   -0.714320  196.718360   
25%        NaN       NaN   -0.493363 -101.953780   -0.677063  199.074557   
50%        NaN       NaN   -0.004680  -99.940945   -0.651600  199.754340   
75%        NaN       NaN    0.466933  -97.814628   -0.630737  200.435847   
max        NaN       NaN    2.227130  -89.903210   -0.576050  203.175010   

                x6                x7          x8          x9         x10  \
count   998.000000               998  998.000000  998.000000  998.000000   
unique         NaN                 7         NaN         NaN         NaN   
top            NaN  Slängpolskorgris         NaN         NaN         NaN   
freq           NaN               359         NaN         NaN         NaN   
mean    -90.445589               NaN   -0.006893   -0.034184   11.015224   
std       1.821704               NaN    1.211136    1.228444    1.351808   
min     -96.716280               NaN   -5.396060   -5.341300    6.138430   
25%     -91.619992               NaN   -0.707320   -0.738320   10.219328   
50%     -90.342965               NaN   -0.021385   -0.030925   11.055555   
75%     -89.232978               NaN    0.656855    0.731315   11.878250   
max     -84.031960               NaN    5.318710    5.673230   15.176510   

               x11   x12         x13  
count   998.000000   998  998.000000  
unique         NaN     2         NaN  
top            NaN  True         NaN  
freq           NaN   997         NaN  
mean     -0.395676   NaN  449.905938  
std       2.216492   NaN    4.740663  
min      -6.576440   NaN  436.182140  
25%      -2.022160   NaN  446.693820  
50%      -0.823380   NaN  449.807960  
75%       1.298115   NaN  452.922118  
max       6.069150   NaN  468.422060

We can see that the x12 feature is irrelevant, because there are 997 true values and 1 false value. We can get rid of feature, because it does not add any information.

In [9]:
df = df_nonnull
df = df.drop('x12', axis = 1)
df.describe(include="all")

y        x1          x2          x3          x4          x5  \
count      998       998  998.000000  998.000000  998.000000  998.000000   
unique       6       996         NaN         NaN         NaN         NaN   
top     Boborg  99.88558         NaN         NaN         NaN         NaN   
freq       420         2         NaN         NaN         NaN         NaN   
mean       NaN       NaN   -0.007644  -99.946945   -0.653185  199.769720   
std        NaN       NaN    0.730013    3.105700    0.031579    1.003122   
min        NaN       NaN   -2.457140 -111.018930   -0.714320  196.718360   
25%        NaN       NaN   -0.493363 -101.953780   -0.677063  199.074557   
50%        NaN       NaN   -0.004680  -99.940945   -0.651600  199.754340   
75%        NaN       NaN    0.466933  -97.814628   -0.630737  200.435847   
max        NaN       NaN    2.227130  -89.903210   -0.576050  203.175010   

                x6                x7          x8          x9         x10  \
count   998.000000               998  998.000000  998.000000  998.000000   
unique         NaN                 7         NaN         NaN         NaN   
top            NaN  Slängpolskorgris         NaN         NaN         NaN   
freq           NaN               359         NaN         NaN         NaN   
mean    -90.445589               NaN   -0.006893   -0.034184   11.015224   
std       1.821704               NaN    1.211136    1.228444    1.351808   
min     -96.716280               NaN   -5.396060   -5.341300    6.138430   
25%     -91.619992               NaN   -0.707320   -0.738320   10.219328   
50%     -90.342965               NaN   -0.021385   -0.030925   11.055555   
75%     -89.232978               NaN    0.656855    0.731315   11.878250   
max     -84.031960               NaN    5.318710    5.673230   15.176510   

               x11         x13  
count   998.000000  998.000000  
unique         NaN         NaN  
top            NaN         NaN  
freq           NaN         NaN  
mean     -0.395676  449.905938  
std       2.216492    4.740663  
min      -6.576440  436.182140  
25%      -2.022160  446.693820  
50%      -0.823380  449.807960  
75%       1.298115  452.922118  
max       6.069150  468.422060

We can see that there is a problem with the features x1. Let's fix it.

In [10]:
df['x1'].describe()

count          998
unique         996
top       99.88558
freq             2
Name: x1, dtype: object

Let's transform the type of x1 from object to float64.

In [11]:
is_valid_entry = pd.to_numeric(df['x1'], errors="coerce").notnull()
df_cleaned = df[is_valid_entry].copy()

df_cleaned['x1'] = pd.to_numeric(df_cleaned['x1'])
df_cleaned['x1'].describe()

count    9.980000e+02
mean     1.002004e+10
std      3.165445e+11
min      9.694035e+01
25%      9.929138e+01
50%      9.996169e+01
75%      1.006510e+02
max      1.000000e+13
Name: x1, dtype: float64

Let's look at the feature x7 since it is a categorical feature.

In [12]:
df_cleaned.x7.unique()

array(['Slängpolskorgris', 'Hambogris', 'Schottisgris', 'Polkagris',
       'Polskorgris', 'Polkagriss', 'Schottisgriss'], dtype=object)

In [13]:
df_cleaned.describe()

x1          x2          x3          x4          x5  \
count  9.980000e+02  998.000000  998.000000  998.000000  998.000000   
mean   1.002004e+10   -0.007644  -99.946945   -0.653185  199.769720   
std    3.165445e+11    0.730013    3.105700    0.031579    1.003122   
min    9.694035e+01   -2.457140 -111.018930   -0.714320  196.718360   
25%    9.929138e+01   -0.493363 -101.953780   -0.677063  199.074557   
50%    9.996169e+01   -0.004680  -99.940945   -0.651600  199.754340   
75%    1.006510e+02    0.466933  -97.814628   -0.630737  200.435847   
max    1.000000e+13    2.227130  -89.903210   -0.576050  203.175010   

               x6          x8          x9         x10         x11         x13  
count  998.000000  998.000000  998.000000  998.000000  998.000000  998.000000  
mean   -90.445589   -0.006893   -0.034184   11.015224   -0.395676  449.905938  
std      1.821704    1.211136    1.228444    1.351808    2.216492    4.740663  
min    -96.716280   -5.396060   -5.341300    6.138430   -6.576440  436.182140  
25%    -91.619992   -0.707320   -0.738320   10.219328   -2.022160  446.693820  
50%    -90.342965   -0.021385   -0.030925   11.055555   -0.823380  449.807960  
75%    -89.232978    0.656855    0.731315   11.878250    1.298115  452.922118  
max    -84.031960    5.318710    5.673230   15.176510    6.069150  468.422060

There are outliers in the feature x1, the maximum value is far too big and totally destroys the mean and the standard deviation. The other features look ok

In [14]:
z_score_treshold = 3
column_name = 'x1'

z_scores = np.abs((df_cleaned[column_name] - df_cleaned[column_name].mean()) / df_cleaned[column_name].std())
outlier_mask = z_scores > z_score_treshold
df_cleaned = df_cleaned[~outlier_mask]
df_cleaned.describe()

x1          x2          x3          x4          x5          x6  \
count  997.000000  997.000000  997.000000  997.000000  997.000000  997.000000   
mean    99.975485   -0.008504  -99.947743   -0.653252  199.769344  -90.445184   
std      1.003027    0.729874    3.107156    0.031524    1.003555    1.822574   
min     96.940350   -2.457140 -111.018930   -0.714320  196.718360  -96.716280   
25%     99.291020   -0.494020 -101.955580   -0.677080  199.074310  -91.623020   
50%     99.960870   -0.006020  -99.942710   -0.651730  199.753330  -90.342270   
75%    100.650860    0.464420  -97.814260   -0.630880  200.435870  -89.228280   
max    103.377680    2.227130  -89.903210   -0.576050  203.175010  -84.031960   

               x8          x9         x10         x11         x13  
count  997.000000  997.000000  997.000000  997.000000  997.000000  
mean    -0.006518   -0.033807   11.013447   -0.401637  449.903556  
std      1.211686    1.229003    1.351319    2.209584    4.742444  
min     -5.396060   -5.341300    6.138430   -6.576440  436.182140  
25%     -0.707970   -0.739150   10.218400   -2.025070  446.689140  
50%     -0.017980   -0.030300   11.055080   -0.824050  449.801480  
75%      0.657040    0.732730   11.876390    1.272990  452.924720  
max      5.318710    5.673230   15.176510    6.069150  468.422060

We have our data now cleaned and without outliers !

# Data modeling

In [15]:
from sklearn.utils import shuffle

In [16]:
feature_columns = ['x1','x2','x3','x4','x5','x6','x7','x8','x9','x10','x11','x13']

X = np.array(df_cleaned[feature_columns].values)
Y = np.array(df_cleaned['y'].values)

Let's correct the categorical feature and the class labels.

In [17]:
classes,indexes,counts = np.unique(Y,return_counts=True,return_index=True)
Nclasses = np.size(classes)

print("{} different classes \n".format(Nclasses))
print(dict(zip(classes,counts)))

6 different classes 

{'Atsutobob': 316, 'Atsutoob': 1, 'Bobborg': 1, 'Boborg': 420, 'Jorggsuto': 1, 'Jorgsuto': 258}


We can see that there are a few mistakes, let's correct them !

In [18]:
Y[indexes[1]] = 'Atsutobob'
Y[indexes[2]] = 'Boborg'
Y[indexes[4]] = 'Jorgsuto'

classes,indexes,counts = np.unique(Y,return_counts=True,return_index=True)
Nclasses = np.size(classes)

print("{} different classes \n".format(Nclasses))
print(dict(zip(classes,counts)))


3 different classes 

{'Atsutobob': 317, 'Boborg': 421, 'Jorgsuto': 259}


They are corrected, now let's correct feature x7.

In [19]:
category,indexes,counts = np.unique(X[:,6],return_counts=True,return_index=True)
Nclasses = np.size(category)

print("{} different classes \n".format(Nclasses))
print(dict(zip(category,counts)))
print(indexes)

7 different classes 

{'Hambogris': 165, 'Polkagris': 316, 'Polkagriss': 1, 'Polskorgris': 22, 'Schottisgris': 133, 'Schottisgriss': 1, 'Slängpolskorgris': 359}
[  1  10 854 133   5 875   0]


In [20]:
x7 = X[:,6]

x7[x7 == 'Polkagriss'] = 'Polkagris'
x7[x7 == 'Polskorgris'] = 'Polkagris'

x7[x7 == 'Schottisgriss'] = 'Schottisgris'

X[:,6] = x7

category,indexes,counts = np.unique(X[:,6],return_counts=True,return_index=True)
Nclasses = np.size(category)

print("{} different classes \n".format(Nclasses))
print(dict(zip(category,counts)))

4 different classes 

{'Hambogris': 165, 'Polkagris': 339, 'Schottisgris': 134, 'Slängpolskorgris': 359}


Let's go back to the datafile structure.

In [21]:
df_cleaned['x7'] = x7
df_cleaned['y'] = Y

In [22]:
RANDOM_STATE = 6

X = df_cleaned[feature_columns]
Y = df_cleaned.y
X_train, y_train = shuffle(X, Y, random_state=RANDOM_STATE)

We need to separate the numerical features from the categorical features (x7), and then we will transform the categorical features into numerical features.

In [23]:
categorical_features = X.select_dtypes(include=['object', 'bool']).columns
numerical_features = X.select_dtypes(include=['float64']).columns

Next, we are going to apply PCA to the numerical features, and we are going to encode the categorical features, in order to make them numerical. For that, we will use a Pipeline from sklearn library. 

In [24]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline

#We encode the categorical features

categorical_transformer = Pipeline([
    ('encoder', OneHotEncoder())
])

In [29]:
from sklearn.decomposition import PCA

#We use a PCA on the numerical features 

numerical_transformer = Pipeline([
    ('pca', PCA(n_components=11)),
])

In [30]:
from sklearn.compose import ColumnTransformer

#We proceed to the columns transformation

preprocessor = ColumnTransformer([
    ('num', numerical_transformer, numerical_features),
    ('cat', categorical_transformer, categorical_features)
])

In [31]:
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, BaggingClassifier, \
GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.linear_model import RidgeClassifierCV
from sklearn.model_selection import cross_val_score, StratifiedKFold
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

classifiers = {
    "Gradient Boosting": GradientBoostingClassifier(random_state=RANDOM_STATE),
    "K-neighbours": KNeighborsClassifier(),
    "Decision tree": DecisionTreeClassifier(),
    "SVM (rbf)": SVC(),
    "SVM (linear)": SVC(kernel="linear"),
    "SVM (polynomial)": SVC(kernel="poly"),
    "Random forest": RandomForestClassifier(random_state=RANDOM_STATE),
    "Adaboost": AdaBoostClassifier(),
    "Bagging": BaggingClassifier(random_state=RANDOM_STATE),
    "MLP": MLPClassifier(max_iter=5000, hidden_layer_sizes=(13,20,20,10,3)),
    "Ridge Classifier": RidgeClassifierCV()
}

best_classifier_name = ""
best_classifier = None
best_score = 0
cross_validation = StratifiedKFold(shuffle=True, random_state=RANDOM_STATE, n_splits=10)

for classifier_name, classifier in classifiers.items():
    
    pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                               ('classifier', classifier)])
    classifier_score = np.average(cross_val_score(pipeline, X_train, y_train, cv=cross_validation))

    print(f"{classifier_name}'s score : {classifier_score}")
    
    if classifier_score> best_score:
        best_score = classifier_score
        best_classifier = pipeline
        best_classifier_name = classifier_name
        
print(f"\n Best classifier : {best_classifier_name}")

Gradient Boosting's score : 0.8786464646464646
K-neighbours's score : 0.660989898989899
Decision tree's score : 0.8125252525252525
SVM (rbf)'s score : 0.7131616161616161
SVM (linear)'s score : 0.7001515151515152
SVM (polynomial)'s score : 0.6639595959595959
Random forest's score : 0.8635252525252526


/opt/anaconda3/lib/python3.12/site-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
/opt/ana

Adaboost's score : 0.7923131313131313
Bagging's score : 0.8555656565656564
MLP's score : 0.6510404040404041
Ridge Classifier's score : 0.686050505050505

 Best classifier : Gradient Boosting


We can see from this analysis that the best classifier is the Gradient Boosting Classifier. Let's use this one to compute the class labels. But before, we will tune the parameters of this gradient boosting classifier, and those of the random forest, since it is the second classifier with the most accuracy.

Let's first tune the parameters of the random forest.

In [32]:
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV

pipeline = Pipeline([('preprocessor', preprocessor), ('forest', RandomForestClassifier(random_state=RANDOM_STATE))])

params = { 
 'forest__bootstrap': [True, False],
 'forest__max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100],
 'forest__max_features': ['log2', 'sqrt'],
 'forest__min_samples_leaf': [1, 2, 4],
 'forest__min_samples_split': [2, 5, 10],
 'forest__n_estimators': [100, 200, 400, 600, 1000, 1200, 1400, 1600, 1800, 2000, 2200, 2400]
}

#Since we can look for all the parameters, because it would be too long, we use RandomizedSearchCV in order to randomly select 120 parameters combinaisons

forest_search = RandomizedSearchCV(pipeline, param_distributions=params, n_iter=120, verbose=1, n_jobs=-1, cv=cross_validation)
forest_search.fit(X_train, y_train)
forest_search.best_score_

Fitting 10 folds for each of 120 candidates, totalling 1200 fits


0.8745858585858585

Now, we tune the parameters for the Gradient Boosting Classifier, and we compare !

In [33]:
pipeline = Pipeline([('preprocessor', preprocessor),('gb', GradientBoostingClassifier(random_state=RANDOM_STATE))])
params = {
    "gb__n_estimators" : [50, 100, 200, 300],
    "gb__max_depth" : [5, 10, 20, None],
    "gb__learning_rate" : [0.1, 0.2, 0.01, 0.05],
}

gb_grid = GridSearchCV(pipeline, param_grid=params, verbose=1, n_jobs=-1, cv=cross_validation)
gb_grid.fit(X_train, y_train)
gb_grid.best_score_

Fitting 10 folds for each of 64 candidates, totalling 640 fits


0.8766262626262626

Ok, from this analysis, we can conclude that both classifiers almost have the same accuracy !
87.4% is pretty good, and I will use the random forest classifier since I am more confortable with this one (this is really a personnal choice and the reason is purely subjective)

Let's look at the parameters : 

In [34]:
forest_search.best_params_

{'forest__n_estimators': 1000,
 'forest__min_samples_split': 10,
 'forest__min_samples_leaf': 1,
 'forest__max_features': 'sqrt',
 'forest__max_depth': 40,
 'forest__bootstrap': True}

# Final Model and Test Labels prediction

In [35]:
estimator = forest_search.best_estimator_

estimator.fit(X, Y)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('pca',
                                                                   PCA(n_components=11))]),
                                                  Index(['x1', 'x2', 'x3', 'x4', 'x5', 'x6', 'x8', 'x9', 'x10', 'x11', 'x13'], dtype='object')),
                                                 ('cat',
                                                  Pipeline(steps=[('encoder',
                                                                   OrdinalEncoder())]),
                                                  Index(['x7'], dtype='object'))])),
                ('forest',
                 RandomForestClassifier(max_depth=40, min_samples_split=10,
                                        n_estimators=1000, random_state=6))])

We need to load the evaluation data.

In [36]:
df_eval = pd.read_csv(EVALUATE_FILE, index_col=0)
df_eval = df_eval.drop('x12',axis = 1) #We get rid of the x12 feature because it is useless.
df_eval.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10000 entries, 1000 to 10999
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x1      10000 non-null  float64
 1   x2      10000 non-null  float64
 2   x3      10000 non-null  float64
 3   x4      10000 non-null  float64
 4   x5      10000 non-null  float64
 5   x6      10000 non-null  float64
 6   x7      10000 non-null  object 
 7   x8      10000 non-null  float64
 8   x9      10000 non-null  float64
 9   x10     10000 non-null  float64
 10  x11     10000 non-null  float64
 11  x13     10000 non-null  float64
dtypes: float64(11), object(1)
memory usage: 1015.6+ KB


In [37]:
x7 = df_eval['x7']
x7[x7=='Polskorgris'] = 'Polkagris'
df_eval['x7'] = x7

/var/folders/b_/bjfw6fm51mn88dk24_v13wsc0000gn/T/ipykernel_36562/2956844396.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x7[x7=='Polskorgris'] = 'Polkagris'


In [38]:
y_predictions = estimator.predict(df_eval)
np.unique(y_predictions)


array(['Atsutobob', 'Boborg', 'Jorgsuto'], dtype=object)

In [39]:
from os import linesep

OUT_FILE = 'data/labels.txt'

with open(OUT_FILE, 'w') as f:
    for y_prediction in y_predictions:
        f.write(y_prediction + linesep)